### PR TITLE
update forbidden messages

### PR DESCRIPTION
**Pull request by deads2k**
Fixes https://github.com/openshift/origin/issues/1558.

Allows for forbidden messages on a verb/resource basis.

@liggitt

### CODE REVIEW

File: pkg/authorization/authorizer/interfaces.go [hunk 1]
```diff
@@ -29,3 +30,15 @@ type AuthorizationAttributes interface {
 	// GetURL returns the URL path being requested, including the leading '/'
 	GetURL() string
 }
+
+// ForbiddenMessageMaker creates a forbidden message from a MessageContext
+type ForbiddenMessageMaker interface {
+	MakeMessage(ctx MessageContext) (string, error)
+}
+
+// MessageContext contains sufficient information to create a forbidden message.  It is bundled in this one object to make it easy and obvious hwo to build a golang template
```
Comment (liggitt) [inline pkg/authorization/authorizer/interfaces.go]: s/hwo/how/

File: pkg/authorization/authorizer/messages.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+)
+
+const DefaultProjectRequestForbidden = "You may not request a new project via this API."
+
+type ForbiddenMessageResolver struct {
+	// TODO if these maps were map[string]map[util.StringSet]ForbiddenMessageMaker, we'd be able to handle cases where sets of resources wanted slightly different messages
+	// unfortunately, maps don't support keys like that, requiring StringSet serialization and deserialization.
+	namespacedVerbsToResourcesToForbiddenMessageMaker map[string]map[string]ForbiddenMessageMaker
```
Comment (smarterclayton) [inline pkg/authorization/authorizer/messages.go]: This wins an award for Longest Variable Name in a Go struct, 2015.

**Comment by deads2k** [comment]
Fixed typo.

**Comment by deads2k** [comment]
@liggitt bump.

File: pkg/authorization/authorizer/messages.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+	return specificMessage, nil
+}
+
+type templateForbiddenMessageMaker struct {
+	parsedTemplate *template.Template
+}
+
+func newTemplateForbiddenMessageMaker(text string) templateForbiddenMessageMaker {
+	parsedTemplate, err := template.New("").Parse(text)
```
Comment (liggitt) [inline pkg/authorization/authorizer/messages.go]: `template.Must(template.New(...`

**Comment by liggitt** [comment]
LGTM with a cursory reading

**Comment by deads2k** [comment]
[merge]

**Comment by deads2k** [comment]
re[merge]

**Comment by deads2k** [comment]
flake:
```
fatal: Unable to create '/data/src/github.com/openshift/origin-bare/refs/heads/master.lock': File exists.
```

re[merge]

**Comment by deads2k** [comment]
flake I haven't seen before:  https://ci.openshift.redhat.com/jenkins/job/merge_pull_requests_origin/2031/
```
error: timed out waiting for the condition
!!! Error in hack/test-cmd.sh:550
  'osc delete all --all' exited with status 1
Call stack:
  1: hack/test-cmd.sh:550 main(...)
Exiting with status 1
```

It passed on the test run: https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/2306/
```
delete all: ok
```


@derekwaynecarr you seen this one before?

[merge] please?

**Comment by deads2k** [comment]
UI failure: https://ci.openshift.redhat.com/jenkins/job/merge_pull_requests_origin/2036/
```
/data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/selenium-webdriver/lib/atoms/error.js:113
  var template = new Error(this.message);
                 ^
UnknownError: Error communicating with the remote browser. It may have died.
```

This thing is cursed.  Pretty please [merge]?

**Pull request merged**